### ISSUE TITLE
Labelled expansion/contraction propagation

### ISSUE THREAD

**Issue by jamievicary** [homotopy-io/homotopy-rs#234]
Today we (NH, LH, CT, JV) discussed how to handle expansion & contraction propagation through slices, in the labelled world. There are four scenarios (see https://arxiv.org/abs/1902.03831 section 5.1):

 - [x]  _Contraction in a singular slice._ Here Lukas had the cool idea that maybe we just feed the entire initial cospan into the labelled contraction algorithm, producing precisely the data needed for the final "composite". This implies that "labelled composition is just contraction". So this is easy (given the existence of labelled contraction.)
 - [x]  _Contraction in a regular slice._ The idea is to bubble as usual, and it seems that we can find the necessary regular slice data canonically. So this is easy.
 - [ ] _Expansion in a regular slice._ Impossible, as with the unlabelled tool.
 - [ ] _Expansion in a singular slice, factorization._ One possibility is to project to the underlying unlabelled expansion problem, factorize as per the unlabelled tool, then perform label propagation #235 to infer the labels on the final geometry. A better possibility might be develop a labelled factorization operation #242.
 - [x] _Expansion in a singular slice, bubbling._ It seems this can work just as before, with the regular data easy to describe.

**Comment by jamievicary** [comment]
Here are some pictures as a useful reference.

![contraction_regular](https://user-images.githubusercontent.com/10602270/143285176-06d2789c-ea8f-4665-9b0b-a2a5ea41d92c.png)
![contraction_singular](https://user-images.githubusercontent.com/10602270/143285178-2268438e-a0a9-43bb-a1cd-af40961b5f67.png)
![expansion_singular](https://user-images.githubusercontent.com/10602270/143285179-99a3f3d1-012a-4772-8983-012b4a014af8.png)

**Comment by NickHu** [comment]
There's an issue with regular propagation, regarding how we currently construct the normalization rewrite (the rewrite that completes the contraction rewrite into a diagram of height 1; i.e. the rewrite which goes from the normalization of the contractum to the contractum).

As we don't yet have normalization worked out, we are using the following scheme to generate normalization rewrites: https://github.com/homotopy-io/homotopy-rs/blob/labelled/homotopy-core/src/contraction.rs#L85-L97. It yields the rewrite which (in the reverse direction) removes cospans that have identical forward and backward rewrites (it applies a smoothing rather than bona fide normalization). This is what we want when we e.g. cancel out a 1-morphism with its inverse, so we get something cup/cap shaped. However, looking at @jamievicary's first picture, this smoothes out the cospan $r_{n+1} \xrightarrow{c} C \xleftarrow{c} r_{n+1}$, and makes the whole rewrite the identity, and nothing happens when this contraction is applied.

**Comment by jamievicary** [comment]
Hi Nick, jump onto Zoom to discuss?

**Comment by jamievicary** [comment]
Thanks Nick, can you explain for my benefit what this shallowness check
does?

**Comment by NickHu** [comment]
It checks at the top level that all the triangles in a rewrite commute. Obviously if this is using composition as defined, this won't work without modification.

**Comment by jamievicary** [comment]
Ah, I understand.

For now we could replace it with a check that everything commutes, except
in dimension 0 just return true, ie don't actually check the label. What do
you think?

**Comment by jamievicary** [comment]
Nick can you update the checkboxes.

**Comment by NickHu** [comment]
Now that #497 is done, this should all work now with a couple caveats:
1. the well-formedness checker hasn't yet been fixed, which causes expansion propagation to bubble when it doesn't need to, yielding unsound behaviour in some expansion-in-a-singular-slice cases;
2. any recursive expansion which triggers an antipushout will crash, as this hasn't been made label-aware yet;
3. (I think) a 2D diagram doesn't have horizontal swipes implemented, as in the unlabelled tool they don't make sense; in particular, this means that the identity on a 1-cell composed with its inverse: 
![f-f-inverse](https://user-images.githubusercontent.com/450276/178105003-955ba1a3-f011-4a6d-b7da-ecf24b239636.png)
can't be cancelled out into a cup above a cap directly. Instead, one must manually enter slice Regular 0, perform the contraction, and then unslice. This yields the result:
![f-cup-cap](https://user-images.githubusercontent.com/450276/178105047-54da3af2-e49e-43c5-a3f7-3605c5f39833.png)

I don't know if there is a 'canonical' test for contraction-in-a-singular-slice, but the example I tested it with had the same behaviour of the unlabelled tool. This example was three beads on three wires: take the identity, contract two beads to be on the same level, then enter the singular slice where that contraction occurs and move the third bead to the same level.
Result:
![contraction-in-singular-slice](https://user-images.githubusercontent.com/450276/178105317-7ff9b968-bfdb-44ee-b1b7-9266143f8bd3.png)

**Comment by calintat** [comment]
3 works for me, does it really not work for you?

**Comment by calintat** [comment]
Oh wait I misunderstood. It sounds like you want to do a a contraction in a singular slice but the diagram doesn't have a singular slice (it only has one regular slice).

**Comment by calintat** [comment]
Also here is the problem with smoothing that I mentioned a few days ago.

If you start with two half braids and do a contraction
<img width="238" alt="Screenshot 2022-07-09 at 13 43 31" src="https://user-images.githubusercontent.com/16837293/178106338-e0669649-76da-4b27-a940-82bc0cd93d8f.png">
the labelled version produces two identity wires:
<img width="237" alt="Screenshot 2022-07-09 at 13 43 37" src="https://user-images.githubusercontent.com/16837293/178106370-cdad64c3-e765-46c2-a38a-6cbac2bb496b.png">
whereas we really want the singular braid:
<img width="239" alt="Screenshot 2022-07-09 at 13 44 00" src="https://user-images.githubusercontent.com/16837293/178106380-a6c54d72-633e-4180-a1b1-4f77609668ad.png">

Basically, the smoothing after contraction should only remove things that came from the result of cancelling inverses, not any other homotopies. I will fix this in my `oriented` branch, as we can easily detect if something came from cancelling inverses as it will have orientation 0.

**Comment by jamievicary** [comment]
Hi, thanks for this, not much time to think in detail today, but the
unlabelled tool does have horizontal swipes in regular levels. This is
contraction in a regular level.

**Comment by calintat** [comment]
It seems impossible to trigger a horizontal swipe in a regular level (in both the unlabelled and labelled versions).

**Comment by calintat** [comment]
I checked code and this is not implemented. I made an issue to do this: #567

**Comment by jamievicary** [comment]
Hmm surprising we missed this.

**Comment by NickHu** [comment]
I think we can close this now; the remaining stuff to do is tracked in other issues
